syntax = "proto3";

package distributed_learning;

service DistributedClient {
  // Training model
  rpc Forward(Query) returns (Tensor) {}
  rpc Backward(Tensor) returns (Query) {}

  // Aggregate model
  rpc GetModelState(Empty) returns (ModelState) {}
  rpc SetModelState(ModelState) returns (Empty) {}

  // Quantize model
  rpc GenerateQuantizedModel(Empty) returns (Empty) {}

  // Test model precision
  rpc TestInference(Query) returns (TensorWithMeasures) {}
  rpc TestQuantizedInference(Query) returns (TensorWithMeasures) {}
}

message Query {
  int32 batch_size = 1;
  int32 request_id = 2;
  int32 status = 3;
}

message Tensor {
  bytes tensor = 1;
  bytes label = 2;
  int32 request_id = 3;
}

message ModelState {
  bytes state = 1;
}

message Measures {
  bytes measures = 1;
}

message TensorWithMeasures {
  Tensor tensor = 1;
  Measures measures = 2;
}

message Empty {}
